google/cloud/internal/url_encode.cc

```diff
@@ -24,35 +23,54 @@ namespace internal {
 
 namespace {
 
-using UrlMappings =
-    std::array<std::pair<absl::string_view, absl::string_view>, 25>;
-
-constexpr auto kEncodeMappings = UrlMappings{{
-    {" ", "%20"}, {"\"", "%22"}, {"#", "%23"},  {"$", "%24"}, {"%", "%25"},
-    {"&", "%26"}, {"+", "%2B"},  {",", "%2C"},  {"/", "%2F"}, {":", "%3A"},
-    {";", "%3B"}, {"<", "%3C"},  {"=", "%3D"},  {">", "%3E"}, {"?", "%3F"},
-    {"@", "%40"}, {"[", "%5B"},  {"\\", "%5C"}, {"]", "%5D"}, {"^", "%5E"},
-    {"`", "%60"}, {"{", "%7B"},  {"|", "%7C"},  {"}", "%7D"}, {"\177", "%7F"},
-}};
-
-// Until we require C++17, it isn't worth trying to eliminate the duplication
-// using a constexpr function to produce kDecodeMappings from kEncodeMappings.
-constexpr auto kDecodeMappings = UrlMappings{{
-    {"%20", " "}, {"%22", "\""}, {"%23", "#"},  {"%24", "$"}, {"%25", "%"},
-    {"%26", "&"}, {"%2B", "+"},  {"%2C", ","},  {"%2F", "/"}, {"%3A", ":"},
-    {"%3B", ";"}, {"%3C", "<"},  {"%3D", "="},  {"%3E", ">"}, {"%3F", "?"},
-    {"%40", "@"}, {"%5B", "["},  {"%5C", "\\"}, {"%5D", "]"}, {"%5E", "^"},
-    {"%60", "`"}, {"%7B", "{"},  {"%7C", "|"},  {"%7D", "}"}, {"%7F", "\177"},
-}};
+bool WeNeedToEscape(unsigned char c) {
+  static auto const* const kSet = new std::set<char>{
+      ' ', '\"', '#', '$', '%', '&',  '+', ',', '/', ':', ';', '<',
+      '=', '>',  '?', '@', '[', '\\', ']', '^', '`', '{', '|', '}',
+  };
+  return std::isprint(c) == 0 || kSet->find(c) != kSet->end();
+}
+
+// Returns 0-15 if c is in [0-9A-Fa-f], and -1 otherwise.
+int ParseDigitHex(char c) {
+  if ('0' <= c && c <= '9') return c - '0';
+  if ('A' <= c && c <= 'F') return c - 'A' + 0xA;
+  if ('a' <= c && c <= 'f') return c - 'a' + 0xa;
+  return -1;
+}
```

Comment:
It looks like I forgot to save another comment in the original review (yikes).  It was along the lines of ...

While I'm not sure we'll ever see an execution character set other than ASCII (and I'm not sure what URL encoding would even mean then), I usually suggest trying not to depend on ASCII ranges or encodings.  For example, consider something like ...
```
int ParseDigitHex(char c) {
  const char kDigits[] = "0123456789ABCDEF";
  const char* cp = std::strchr(kDigits, std::toupper(c));
  return (cp == nullptr || *cp == '\0') ? -1 : cp - kDigits;
}
```

PS: s/ParseDigitHex/ParseHexDigit/?

Reply:
hmm. I am relying on the fact that the encoding goes `0123456789` and `ABCDEF`....

Using `strchr` will have to look through all of the characters... idk.

I changed the name only.